design/ceph/object/swift-and-keystone-integration.md

```diff
@@ -0,0 +1,312 @@
+  name: rook-ceph-object-subuser-my-store-my-user:swift [1]
+  namespace: rook-ceph
+data:
+  SWIFT_USER: my-user:swift                             [2]
+  SWIFT_SECRET_KEY: $KEY                                [3]
+```
+Annotations:
+* `[1]` The name is constructed by joining the following elements with dashes
+  (compare [the corresponding name of the secret for the object store users](
+  https://github.com/rook/rook/blob/376ca62f8ad07540d9ddffe9dc0ee53f4ac35e29/pkg/operator/ceph/object/user/controller.go#L416)):
+      - the literal `rook-ceph-object-subuser`
+      - the name of the object store resource
+      - the full name of the subuser (including the `username:`-prefix).
+* `[2]` The full name of the subuser (including the `username:`-prefix).
+* `[3]` The generated swift access secret.
+
+### Risks and Mitigation
+
+As long as the Object Store CRD changes are well thought out the
+overall risk is minimal.
+
+* If Swift is enabled by accident this could lead to an increased and
+  unexpected attack surface, especially since the authentication and
+  authorization for Swift and S3 may differ depending on how Ceph RGW
+  is configured. This is mitigated by requiring explicit configuration
+  to enable Swift.
+
+* A misconfigured Keystone integration may allow users to gain access
+  to objects they should not be authorized to access (e.g. if the
+  `rgw keystone accepted roles` setting is too broad). This will be
+  mitigated by proper documentation to make the operator aware of the
+  risk.
+
+* Ceph RGW allows to disable TLS verification when querying Keystone,
+  we deliberately choose not to expose this config option to the user.
+
+## Drawbacks
+
+* As shown in [#4754](https://github.com/rook/rook/issues/4754)
+  keystone can be integrated via config override so
+  it is not strictly necessary to support configuring it via the
+  Object Store CRD. Adding it to the CRD complicates things with
+  minor gain.
+
+## Alternatives
+
+* There is no workable alternative to extending the CRD for Swift
+  support.
+
+* Keystone support could be configured via Rook config override as shown
+  in [#4754](https://github.com/rook/rook/issues/4754).
+
+## Open Questions
+
+### How to support multiple APIs in one Gateway
+
+Having a single Ceph RGW deployment serve both S3 and Swift is both
+possible and often desirable for interoperability reasons. The current
+structure of the Object Store CRD makes it difficult to represent the
+configuration for this cleanly.
+
+Possible Solutions:
+
+* Run a second object store resource. The question is then how to use
+  the same pools (since the pools are derived from the resource name
+  which must be unique)
+
+* Simply allow a list in the type argument (but keep allowing strings
+  for backwards compatibility). The drawback here is that the gateway
+  config will then be cluttered by the config options pertaining to
+  different API types.
+
+* Break compatibility with the existing CRD and split up the
```

Comment:
For compatibility let's treat any settings as s3-only if they don't apply to swift. Or else I'm not clear why we need a breaking change here.

Reply:
I definitely agree that we shouldn't break backward compatibility. We should assume S3 for anything pre-existing.

Reply:
Noted, the Open Questions have been removed (and the chosen solution is documented in the main document). Backwards compatibility is maintained.